# Context-first lattice analysis

Read the original texts before consulting category labels. This notebook
replays saved evidence and numerical audits, not fresh model inference.
The qualitative choices are exploratory and not analyst-blinded.
See `experiments/activation_studies/contextreading_v1/READINGS.md`.

A join intersects extents; a meet weakens requirements and can admit extra
items. FG closes a numerical description while preserving its extent.
None of these laws by itself assigns semantic meaning.


In [1]:
from pathlib import Path
import json
import os
import sys

root = Path(os.environ.get("LATTICE_REPOSITORY", Path.cwd())).resolve()
while not (root / "src/lattmc").exists() and root != root.parent:
    root = root.parent
if not (root / "src/lattmc").exists():
    raise RuntimeError("Set LATTICE_REPOSITORY to the checkout root.")
sys.path[:0] = [str(root), str(root / "src")]
out = Path(os.environ.get(
    "LATTICE_READING_RESULTS",
    root / "data/activation_studies/contextreading_v1"))
print("Interpreter:", sys.executable)
print("Records:", out)


Interpreter: /Users/ltsinadze/git/my_papers/.venv/bin/python
Records: /Users/ltsinadze/git/my_papers/data/activation_studies/contextreading_v1


In [2]:
from lattmc.contextstudy.semantic_audit_codexgen import audit

print(audit(root, out))


{'query_records': 1686, 'distinct_dense_masks': 942, 'distinct_closed_extents': 619, 'token_witness_records': 157, 'prefix_contrasts': 7}


## The complete operation grid

Every full and selected query is retained, including zero meets and empty
joins. Counts use 25,600 original GPT-2 chunks, including legacy BOS.
Four source rows are refreshed in memory; other cached rows are unchanged.
These chunks are not independent original documents.


In [3]:
import pandas as pd
from IPython.display import HTML, display

rows = []
for kind in ("sae", "tc"):
    for block in (0, 8, 11):
        data = json.loads((out / f"gpt2_{kind}{block}.json").read_text())
        for r in data["records"]:
            rows.append({
                "condition": f"{kind}{block}", "group": r["group"],
                "rule": r["rule"],
                "constituents": [c["count"] for c in r["components"]],
                "meet": r["meet"]["count"], "join": r["join"]["count"],
                "additional meet": r["extra_count"],
                "closed meet coordinates": r["meet"]["closed_positive"],
            })
frame = pd.DataFrame(rows)
print("Complete grid shape:", frame.shape)
display(frame[frame["condition"] == "tc11"])


Complete grid shape: (330, 8)


,condition,group,rule,constituents,meet,join,additional meet,closed meet coordinates
275,tc11,biological_pair,full,"[1, 1]",2,0,0,2245
276,tc11,biological_pair,rank1,"[349, 1010]",25600,232,24473,30
277,tc11,biological_pair,rank2,"[620, 599]",25600,164,24545,30
278,tc11,biological_pair,rank3,"[1407, 2631]",25600,272,21834,30
279,tc11,biological_pair,rank23,"[620, 2631]",25600,31,22380,30
280,tc11,research_three,full,"[1, 1, 1]",99,0,97,208
281,tc11,research_three,rank1,"[134, 385, 9]",25600,1,25094,30
282,tc11,research_three,rank2,"[1752, 1313, 319]",25600,17,22488,30
283,tc11,research_three,rank3,"[530, 792, 532]",25600,1,23822,30
284,tc11,research_three,rank23,"[1752, 792, 319]",25600,5,22948,30


## Original texts and exact witnesses

Rose marks any positive-coordinate witness. Bold rose marks a token that
satisfies the entire query. Yellow D marks collective item satisfaction
without any whole-query token. These are numerical, not semantic labels.
The full original sequence decoding is printed below each token display;
separately decoded UTF-8 byte fragments may otherwise show replacements.
Queries retain recorded amplitudes, with no membership epsilon.


In [4]:
from lattmc.contextstudy.witnessrender_codexgen import snippet

choices = [
    ("tc11", "research_three", "full", "meet", 16621),
    ("tc11", "mixed_pair", "rank1", "join", 9155),
    ("tc11", "biological_four", "rank1", "join", 23609),
    ("tc11", "mixed_three", "rank3", "join", 12902),
    ("tc11", "cat_dog", "full", "meet", 17545),
    ("sae11", "new_de", "rank23", "join", 21747),
]
for condition, group, rule, op, row in choices:
    data = json.loads((out / f"{condition}_witnesses.json").read_text())
    r = dict(next(r for r in data["records"] if r["row"] == row
                  and r["group"] == group and r["rule"] == rule
                  and r["operation"] == op))
    r["valid_positions"] = list(range(128))
    r["diagnostic"] = (r["partial"] or [0])[0]
    badge = r["status"]
    if badge == "D":
        badge = '<mark style="background:#FFEF9F">D</mark>'
    display(HTML(f'<h4>{condition}: {group}, {rule} {op}, row {row}</h4>'
                 + badge + '<p>' + snippet(r, as_html=True) + '</p>'))
    print("Complete original decode:", r["raw_text"])
    print("Extent:", r["extent"], "Exact query:", r["query"])
    print("Coordinate inequalities:", r["inequalities"])


Complete original decode: <|endoftext|>atter counterpart, positrons. Both types of collisions sometimes result in B mesons, however, allowing each to measure the end products when the unstable mesons decay. In a paper published in the September 11 issue of Physical Review Letters, the LHCb team announced that they had observed a potential excess of taus about 25 to 30 percent greater than the frequency predicted by the Standard Model. Belle saw a similar, but less pronounced, effect, in data reported in a paper under review at Physical Review D. Both teams shared their findings in May at the Flavor Physics & CP Violation 2015 conference in Nagoya, Japan.


Extent: 99 Exact query: {'coordinates': [1066, 2463, 5914, 7172, 8296, 10317, 11424, 12053, 12278, 12476, 13795, 15117, 16100, 18705, 20520, 20581, 21135, 22275], 'values': [1.0813539028167725, 0.4940001964569092, 0.0437740832567215, 0.21120122075080872, 0.8520114421844482, 1.241648554801941, 0.3533915877342224, 0.2785995602607727, 0

Complete original decode: <|endoftext|> while Hall may find himself matched up with Welker some, a lot of the responsibility of covering him likely will fall on Kevin Barnes. The Redskins occasionally have put fellow corner Josh Wilson on the opponent’s slot receiver, however, and put Barnes on the outside. It will take a collective effort to limit Welker.

WSH tackles Sean Locklear/Willie Smith vs. NE defensive end Andre Carter

The Redskins haven’t yet announced Trent Williams’ replacement, but it will be either eight-year veteran Sean Locklear, who started two games this season at that spot with Williams injured, or
Extent: 6 Exact query: {'coordinates': [4355, 13040], 'values': [11.577988624572754, 23.46306037902832]}
Coordinate inequalities: [{'coordinate': 4355, 'requirement': 11.577988624572754, 'item_max': 11.709681510925293, 'cached_max': 11.709688186645508, 'witness_positions': [36], 'witness_pieces': [' Wilson']}, {'coordinate': 13040, 'requirement': 23.46306037902832, 'item

Complete original decode: <|endoftext|> model in worms or flies, whose pathways of starch digestion and steroid sensitivity differ dramatically from those used by mammals. The growing availability of tissue banks from mice treated with these agents should serve as a spur toward increased collaboration on the pathways by which these drugs lead to preserved health and postponed mortality in mice.I love you, crazy ranting nutcase person.

I love you, crazy ranting nutcase person.

Saints be praised, this guy is amazing. Aficionados of the GOP's new move towards abject batshit insanity have found a new poster boy in inexplicable Virginia lt. gov. nominee E
Extent: 3 Exact query: {'coordinates': [3493, 15788, 16100, 16566], 'values': [6.1118550300598145, 8.904752731323242, 5.924178600311279, 4.358260154724121]}
Coordinate inequalities: [{'coordinate': 3493, 'requirement': 6.1118550300598145, 'item_max': 7.227379322052002, 'cached_max': 7.22738790512085, 'witness_positions': [7], 'witness_pi

Complete original decode: <|endoftext|> complain to the Department of Justice that the A.D.A. is too broad in its definition of “service animal,” and too restrictive of businesses trying to protect themselves from people who fake it. Which many people do.

In October, a man in Portland, Ore., took his dog on a bus, claiming that it was a service animal. While getting off the bus, the dog killed another dog that was riding as a “comfort animal.” (In Portland, comfort animals are allowed on public transportation.) A few days later, an editorial appeared in The Oregonian with the headline “
Extent: 1 Exact query: {'coordinates': [4034, 6383, 12780], 'values': [6.6127753257751465, 5.553518295288086, 10.132925033569336]}
Coordinate inequalities: [{'coordinate': 4034, 'requirement': 6.6127753257751465, 'item_max': 7.348630428314209, 'cached_max': 7.348629474639893, 'witness_positions': [4], 'witness_pieces': [' Department']}, {'coordinate': 6383, 'requirement': 5.553518295288086, 'item_max':

Complete original decode: <|endoftext|> consciousness before she hit the flowerbed. Petals flew throughout the air, dusted lightly with dew. A few bottles in her knapsack leaked out, as that one old vine still clung to it. It was ripped off now, the shine of glucose dripping onto the dirt.

It was the first thoughtless sleep she would have in ages.At the American Heartland Forum in Columbia, Missouri before the upcoming presidential primary in the state (which is non-binding and awards zero delegates), Rick Santorum joined Focus on the Family founder James Dobson to push the myth that the recently passed health care reform law would
Extent: 25 Exact query: {'coordinates': [680, 940, 1012, 1199, 1857, 1902, 2357, 2510, 2604, 3251, 3554, 4917, 4937, 5707, 5747, 5934, 6081, 6536, 7328, 8616, 8964, 9151, 9647, 9885, 11193, 12953, 12972, 13068, 14003, 14445, 14576, 14800, 15348, 18044, 19636, 22146, 22746], 'values': [0.15167096257209778, 0.6644694805145264, 0.2926943004131317, 0.2931274473

Complete original decode: <|endoftext|> that an Indian minister would publicly mimic the US’ slanderous accusations against Russia by implying that Moscow is using Twitter bots to support the opposition is worrisome and suggests that pro-American sentiment in New Delhi is even stronger than the most vocal critics imagine it to be.

Andrew Korybko is an American Moscow-based political analyst specializing in the relationship between the US strategy in Afro-Eurasia, China’s One Belt One global vision of New Silk Road connectivity, and Hybrid Warfare.

Featured image is from the author.Alfonso Cuaron Explains Why Robert Downey Jr. Wasn
Extent: 13 Exact query: {'coordinates': [7275, 22858], 'values': [26.5772705078125, 12.592910766601562]}
Coordinate inequalities: [{'coordinate': 7275, 'requirement': 26.5772705078125, 'item_max': 57.14857864379883, 'cached_max': 57.1485710144043, 'witness_positions': [39, 95], 'witness_pieces': [' New', ' New']}, {'coordinate': 22858, 'requirement': 12.592

## Prefix contrasts after discovering a candidate relation

The virus/Clippers join has six members. Reading them suggests a put/place
complement role combined with sports-team names. Other-verb substitutions
make coordinate 4355 zero at seven source/returned targets. Subsequent
family alternatives preserve positivity, but only one satisfies the
original threshold. This is adaptive input sensitivity, not latent ablation
or feature-mediated downstream behavior. Suffix controls are unchanged.


In [5]:
data = json.loads((out / "prefix_contrast.json").read_text())
columns = ["row", "target", "before", "after", "original", "edited",
           "family_replacement", "family_activation",
           "full_target_code_suffix_equal"]
display(pd.DataFrame(data["records"])[columns])


,row,target,before,after,original,edited,family_replacement,family_activation,full_target_code_suffix_equal
0,3457,65,put,introduced,11.577989,0.0,placed,6.096639,True
1,5342,108,put,get,22.152746,0.0,place,10.878114,True
2,9155,36,put,used,11.709682,0.0,placed,9.244247,True
3,9486,22,placing,spending,11.825215,0.0,putting,15.271898,True
4,10441,120,puts,leaves,15.240685,0.0,places,7.853972,True
5,14210,55,put,apply,12.666020,0.0,place,5.018703,True
6,25220,117,put,score,18.828161,0.0,place,2.204954,True


## Structural readings before the later label comparison

Pythia's early-period pattern includes abbreviation periods, not just title
separators. SmolLM2's name-fragment pattern has a year-digit counterexample.
Members and nonmembers were sampled without new category selection. Earlier
source labels were known; this is not a blinded semantic precision study.


In [6]:
data = json.loads((out / "structure_probe.json").read_text())
for family, row in [("pythia_topk", 1723), ("smol_topk", 1810)]:
    r = dict(next(r for r in data["records"] if
                  r["family"] == family and r["row"] == row))
    r.update(valid_positions=r["valid"], partial=r["whole"], status="S")
    r["diagnostic"] = r["whole"][0]
    display(HTML(f'<h4>{family}, row {row}</h4><p>'
                 + snippet(r, as_html=True) + '</p>'))
    print(r["raw_text"])
    print("Whole-query positions:", r["whole"], "Query:", r["query"])
notes = json.loads((out / "interpretations_before_labels.json").read_text())
print("Recorded text-first interpretations:")
print(json.dumps(notes, indent=2, ensure_ascii=False))


St. John's Church Oakfield Ryde.  St. John's Church Oakfield Ryde is a parish church in the Church of England located in Ryde Isle of Wight.
Whole-query positions: [2] Query: {'coordinates': [1512], 'values': [57.30409240722656]}


Karaağaç Koçarlı.  Karaağaç is a village in the District of Koçarlı Aydın Province Turkey. As of 2010 it had a population of 164 people.
Whole-query positions: [39, 40, 41] Query: {'coordinates': [3328], 'values': [31.251203536987305]}
Recorded text-first interpretations:
{
  "analysis_status": "Single model-assisted close reading; exploratory, not independent gold labels.",
  "pythia": "All sixteen members have whole-query witnesses on the first full stop. Fourteen stops are the title separator, two abbreviate St. within the title. The interpretation is early-period activity, not necessarily a title-boundary feature.",
  "smol": "Witnesses repeatedly occur on proper-name fragments and romanizations; row1810 instead satisfies the query only at a space and digits in 2010. A universal naming interpretation is unsupported.",
  "mixed_pair": "All six joined texts discuss sports and a put/place construction. Coordinate4355 witnesses are their complements, while13040 witnesses name sports te

In [7]:
from lattmc.contextstudy.operations_codexgen import digest

labels = json.loads((out / "labels_after_reading.json").read_text())
assert labels["interpretation_sha256"] == digest(
    out / "interpretations_before_labels.json")
print("Subsequent category comparison:")
print(json.dumps(labels["families"], indent=2))


Subsequent category comparison:
{
  "pythia_topk": {
    "member_label_counts": {
      "0": 3,
      "2": 1,
      "4": 2,
      "6": 1,
      "7": 2,
      "8": 1,
      "10": 1,
      "12": 2,
      "13": 3
    },
    "distinct_labels": 9,
    "naturalplace_count": 2,
    "sample_size": 16
  },
  "smol_topk": {
    "member_label_counts": {
      "0": 2,
      "1": 1,
      "3": 1,
      "4": 2,
      "5": 1,
      "8": 4,
      "9": 1,
      "10": 1,
      "11": 1,
      "12": 2
    },
    "distinct_labels": 10,
    "naturalplace_count": 0,
    "sample_size": 16
  }
}


## Reproduction and limitations

The source modules and staged commands are documented in the study README.
`exploration_codexgen` computes cached extents and closure;
`readingreplay_codexgen` runs bounded GPT-2 replay;
`prefixcontrast_codexgen` runs the input contrasts;
`structureprobe_codexgen` replays the fixed original MPS batches;
`semantictables_codexgen` regenerates the unified TeX tables.

No latent ablation, reconstruction-controlled behavioral test, independent
contextual annotation, or new corpus was added. Category mismatches do not
rule out contextual structure; exact numerical membership does not prove
that every returned item has a coherent semantic interpretation.
